apps/conversations/models.py

```diff
@@ -157,6 +163,31 @@ def was_edited(self):
     def is_deleted(self):
         return self.deleted_at is not None
 
+    def add_reaction(self, user, emoji_shortcode):
+        """Adds a reaction by `user` to this message.
+
+        This method is idempotent. If it is called multiple times with the same arguments, it will only create 1 record of the emoji reaction.
+        """
+
+        self.emoji_reactions.get_or_create(
+            user=user,
+            emoji_shortcode=emoji_shortcode,
+        )
+
+    def remove_reaction(self, user, emoji_shortcode):
+        """Removes an existing reaction by `user` to this message.
+
+        This method is idempotent. If it is called multiple times with the same arguments, it will remove the record of the emoji reaction.
+
+        If it is called multiple times, and the reaction has already been deleted, this method has no effect and just performs a no-op.
+        """
+        reaction = self.emoji_reactions.filter(
+            user=user,
+            emoji_shortcode=emoji_shortcode,
+        ).first()
+        if reaction:
+            reaction.delete()
```

Comment:
```suggestion
        reaction = self.emoji_reactions.filter(
            user=user,
            emoji_shortcode=emoji_shortcode,
        ).delete()
```

See: https://docs.djangoproject.com/en/5.0/topics/db/queries/#deleting-objects
> You can also delete objects in bulk. Every [QuerySet](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#django.db.models.query.QuerySet) has a [delete()](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#django.db.models.query.QuerySet.delete) method, which deletes all members of that [QuerySet](https://docs.djangoproject.com/en/5.0/ref/models/querysets/#django.db.models.query.QuerySet).

Also: https://docs.djangoproject.com/en/5.0/ref/models/querysets/#delete
> The delete() method does a bulk delete and does not call any delete() methods on your models. It does, however, emit the [pre_delete](https://docs.djangoproject.com/en/5.0/ref/signals/#django.db.models.signals.pre_delete) and [post_delete](https://docs.djangoproject.com/en/5.0/ref/signals/#django.db.models.signals.post_delete) signals for all deleted objects (including cascaded deletions).

Depending on the implementation/needs, we might want to hook up with `post_delete` to update clients who have that emoji status.

Reply:
Will look through the docs to understand things better.